onnxruntime/core/providers/vitisai/imp/ep_context_utils.cc

```diff
@@ -0,0 +1,831 @@
+    output_node_arg_ptrs.push_back(&temp_node_arg);
+  }
+
+  // Attr "embed_mode".
+  auto p_attr_0 = ONNX_NAMESPACE::AttributeProto::Create();
+  p_attr_0->set_name(kEmbedModeAttr);
+  // p_attr_0->set_type(onnx::AttributeProto_AttributeType_INT);
+  p_attr_0->set_type(ONNX_NAMESPACE::AttributeProto::INT);
+  p_attr_0->set_i(embed_mode);
+  // Attr "ep_cache_context".
+  auto p_attr_1 = ONNX_NAMESPACE::AttributeProto::Create();
+  p_attr_1->set_name(kEPCacheContextAttr);
+  // p_attr_1->set_type(onnx::AttributeProto_AttributeType_STRING);
+  p_attr_1->set_type(ONNX_NAMESPACE::AttributeProto::STRING);
+  // Relative to the ONNX model file.
+  p_attr_1->set_s(
+      embed_mode == 0 ? fs::path(ctx_cache_file_loc).filename().string() : serialized_ctx_cache);
+  // Attr "source".
+  auto p_attr_2 = ONNX_NAMESPACE::AttributeProto::Create();
+  p_attr_2->set_name(kSourceAttr);
+  // p_attr_2->set_type(onnx::AttributeProto_AttributeType_STRING);
+  p_attr_2->set_type(ONNX_NAMESPACE::AttributeProto::STRING);
+  p_attr_2->set_s(kVitisAIExecutionProvider);
+  // Attr "onnx_model_filename".
+  auto p_attr_3 = ONNX_NAMESPACE::AttributeProto::Create();
+  p_attr_3->set_name(kONNXModelFileNameAttr);
+  // p_attr_3->set_type(onnx::AttributeProto_AttributeType_STRING);
+  p_attr_3->set_type(ONNX_NAMESPACE::AttributeProto::STRING);
+  p_attr_3->set_s(fs::path(graph_viewer.ModelPath().ToPathString()).filename().string());
+  // Attr "notes".
+  auto p_attr_4 = ONNX_NAMESPACE::AttributeProto::Create();
+  p_attr_4->set_name(kNotesAttr);
+  // p_attr_4->set_type(onnx::AttributeProto_AttributeType_STRING);
+  p_attr_4->set_type(ONNX_NAMESPACE::AttributeProto::STRING);
+  // FIXME: 2G-limit of ProtoBuf.
+  if (saving_orig_graph) {
+    p_attr_4->set_s(SerializeOrigialGraph(graph_viewer));
+    LOGS_DEFAULT(VERBOSE) << "Saved serialized graph to attr proto";
+  } else {
+    nlohmann::json j_obj;
+    j_obj["backend_cache_dir"] = backend_cache_dir;
+    j_obj["backend_cache_key"] = backend_cache_key;
+    p_attr_4->set_s(j_obj.dump(-1, ' ', false, nlohmann::json::error_handler_t::replace));
+    LOGS_DEFAULT(VERBOSE) << "Saved backend cache key to attr proto";
+  }
+
+  auto p_node_attrs = NodeAttributes::Create();
+  constexpr int num_attrs = 5;
+  p_node_attrs->reserve(num_attrs);
+  p_node_attrs->emplace(kEmbedModeAttr, *p_attr_0);
+  p_node_attrs->emplace(kEPCacheContextAttr, *p_attr_1);
+  p_node_attrs->emplace(kSourceAttr, *p_attr_2);
+  p_node_attrs->emplace(kONNXModelFileNameAttr, *p_attr_3);
+  p_node_attrs->emplace(kNotesAttr, *p_attr_4);
+
+  // Since we don't implement `IExecutionProvider::GetEpContextNodes()` and
+  // thus don't leverage `CreateEpContextModel()` in the file "graph_partitioner.cc",
+  // we specify a brand-new node name here.
+  ep_ctx_graph.AddNode(kEPContextOpName, kEPContextOp, "", input_node_arg_ptrs, output_node_arg_ptrs, p_node_attrs.get(), kEPContextOpDomain);
```

Comment:
The easiest way is to use 
1. auto& ep_node = graph.AddNode(), 
2. ep_node.AddAttribute().
refers to https://github.com/microsoft/onnxruntime/blob/4743803944d6f6333bf93b89a62d8083d4466710/onnxruntime/core/providers/qnn/builder/onnx_ctx_model_helper.cc#L221-L253

Reply:
Yes, I totall agree.
Essentially, the two ways are the same, but one is "top-down" style while the other is "bottom-up" style.

For this iteration, can we keep this?